// @wordpress/block-library/build/social-links/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SocialLinksEdit = SocialLinksEdit;
exports.default = void 0;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var ALLOWED_BLOCKS = ['core/social-link'];
var sizeOptions = [{
  name: (0, _i18n.__)('Small'),
  value: 'has-small-icon-size'
}, {
  name: (0, _i18n.__)('Normal'),
  value: 'has-normal-icon-size'
}, {
  name: (0, _i18n.__)('Large'),
  value: 'has-large-icon-size'
}, {
  name: (0, _i18n.__)('Huge'),
  value: 'has-huge-icon-size'
}];

function SocialLinksEdit(props) {
  var _props$attributes = props.attributes,
      size = _props$attributes.size,
      openInNewTab = _props$attributes.openInNewTab,
      setAttributes = props.setAttributes;
  var SocialPlaceholder = (0, _element.createElement)("div", {
    className: "wp-block-social-links__social-placeholder"
  }, (0, _element.createElement)("div", {
    className: "wp-social-link"
  }), (0, _element.createElement)("div", {
    className: "wp-block-social-links__social-placeholder-icons"
  }, (0, _element.createElement)("div", {
    className: "wp-social-link wp-social-link-twitter"
  }), (0, _element.createElement)("div", {
    className: "wp-social-link wp-social-link-facebook"
  }), (0, _element.createElement)("div", {
    className: "wp-social-link wp-social-link-instagram"
  })));
  var className = (0, _classnames.default)(size);
  var blockProps = (0, _blockEditor.useBlockProps)({
    className: className
  });
  var innerBlocksProps = (0, _blockEditor.__experimentalUseInnerBlocksProps)(blockProps, {
    allowedBlocks: ALLOWED_BLOCKS,
    orientation: 'horizontal',
    placeholder: SocialPlaceholder,
    templateLock: false,
    __experimentalAppenderTagName: 'li'
  });
  var POPOVER_PROPS = {
    position: 'bottom right',
    isAlternate: true
  };
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_components.ToolbarItem, null, function (toggleProps) {
    return (0, _element.createElement)(_components.DropdownMenu, {
      label: (0, _i18n.__)('Size'),
      text: (0, _i18n.__)('Size'),
      icon: null,
      popoverProps: POPOVER_PROPS,
      toggleProps: toggleProps
    }, function (_ref) {
      var onClose = _ref.onClose;
      return (0, _element.createElement)(_components.MenuGroup, null, sizeOptions.map(function (entry) {
        return (0, _element.createElement)(_components.MenuItem, {
          icon: (size === entry.value || !size && entry.value === 'has-normal-icon-size') && _icons.check,
          isSelected: size === entry.value,
          key: entry.value,
          onClick: function onClick() {
            setAttributes({
              size: entry.value
            });
          },
          onClose: onClose,
          role: "menuitemradio"
        }, entry.name);
      }));
    });
  }))), (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Link settings')
  }, (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Open links in new tab'),
    checked: openInNewTab,
    onChange: function onChange() {
      return setAttributes({
        openInNewTab: !openInNewTab
      });
    }
  }))), (0, _element.createElement)("ul", innerBlocksProps));
}

var _default = SocialLinksEdit;
exports.default = _default;
