// @wordpress/block-library/build-module/social-links/edit.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classNames from 'classnames';
/**
 * WordPress dependencies
 */

import { Fragment } from '@wordpress/element';
import { BlockControls, __experimentalUseInnerBlocksProps as useInnerBlocksProps, useBlockProps, InspectorControls } from '@wordpress/block-editor';
import { DropdownMenu, MenuGroup, MenuItem, PanelBody, ToggleControl, ToolbarItem, ToolbarGroup } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons';
var ALLOWED_BLOCKS = ['core/social-link'];
var sizeOptions = [{
  name: __('Small'),
  value: 'has-small-icon-size'
}, {
  name: __('Normal'),
  value: 'has-normal-icon-size'
}, {
  name: __('Large'),
  value: 'has-large-icon-size'
}, {
  name: __('Huge'),
  value: 'has-huge-icon-size'
}];
export function SocialLinksEdit(props) {
  var _props$attributes = props.attributes,
      size = _props$attributes.size,
      openInNewTab = _props$attributes.openInNewTab,
      setAttributes = props.setAttributes;
  var SocialPlaceholder = createElement("div", {
    className: "wp-block-social-links__social-placeholder"
  }, createElement("div", {
    className: "wp-social-link"
  }), createElement("div", {
    className: "wp-block-social-links__social-placeholder-icons"
  }, createElement("div", {
    className: "wp-social-link wp-social-link-twitter"
  }), createElement("div", {
    className: "wp-social-link wp-social-link-facebook"
  }), createElement("div", {
    className: "wp-social-link wp-social-link-instagram"
  })));
  var className = classNames(size);
  var blockProps = useBlockProps({
    className: className
  });
  var innerBlocksProps = useInnerBlocksProps(blockProps, {
    allowedBlocks: ALLOWED_BLOCKS,
    orientation: 'horizontal',
    placeholder: SocialPlaceholder,
    templateLock: false,
    __experimentalAppenderTagName: 'li'
  });
  var POPOVER_PROPS = {
    position: 'bottom right',
    isAlternate: true
  };
  return createElement(Fragment, null, createElement(BlockControls, null, createElement(ToolbarGroup, null, createElement(ToolbarItem, null, function (toggleProps) {
    return createElement(DropdownMenu, {
      label: __('Size'),
      text: __('Size'),
      icon: null,
      popoverProps: POPOVER_PROPS,
      toggleProps: toggleProps
    }, function (_ref) {
      var onClose = _ref.onClose;
      return createElement(MenuGroup, null, sizeOptions.map(function (entry) {
        return createElement(MenuItem, {
          icon: (size === entry.value || !size && entry.value === 'has-normal-icon-size') && check,
          isSelected: size === entry.value,
          key: entry.value,
          onClick: function onClick() {
            setAttributes({
              size: entry.value
            });
          },
          onClose: onClose,
          role: "menuitemradio"
        }, entry.name);
      }));
    });
  }))), createElement(InspectorControls, null, createElement(PanelBody, {
    title: __('Link settings')
  }, createElement(ToggleControl, {
    label: __('Open links in new tab'),
    checked: openInNewTab,
    onChange: function onChange() {
      return setAttributes({
        openInNewTab: !openInNewTab
      });
    }
  }))), createElement("ul", innerBlocksProps));
}
export default SocialLinksEdit;
